ocaml/typing/ident.ml

```diff
@@ -107,28 +146,85 @@ let equal i1 i2 =
   | _ ->
       false
 
+let same_unscoped us1 us2 = (get_desc us1).stamp = (get_desc us2).stamp
+
 let same i1 i2 =
   match i1, i2 with
   | Local { stamp = s1; _ }, Local { stamp = s2; _ }
   | Scoped { stamp = s1; _ }, Scoped { stamp = s2; _ }
   | Predef { stamp = s1; _ }, Predef { stamp = s2 } ->
       s1 = s2
+  | Unscoped us1, Unscoped us2 ->
+    same_unscoped us1 us2
   | Global name1, Global name2 ->
       name1 = name2
   | Instance (f1, args1), Instance (f2, args2) ->
       String.equal f1 f2 && List.equal String.equal args1 args2
   | _ ->
       false
 
+let stamp_us us = (get_desc us).stamp
+
 let stamp = function
   | Local { stamp; _ }
   | Scoped { stamp; _ } -> stamp
+  | Unscoped us -> stamp_us us
   | _ -> 0
 
+type change = unscoped * unscoped_state
+
+let change_log = ref (fun _ -> assert false)
+
+let undo_change (us, us_d) =
+  us.state <- us_d
+
+let rec repr us =
+  match us.state with
+  | Ulink us' ->
+    repr us'
+  | Udesc _ -> us
+
+let link_unscoped us1 us2 =
+  let us1 = repr us1 in
+  let us2 = repr us2 in
+  if us1 == us2 then () else begin
+  !change_log (us1, us1.state);
+  us1.state <- Ulink us2
+  end
+
+let id_pairs = ref []
```

Comment:
Doing this would require `ctype.ml` to give the content of that variable to `Path.equiv` that will afterward give it to `Ident.equiv` .
I put it originally in `ctype.ml` because it was the only place where it was modified but moved it here to avoid all this moving around.